app/lib/actions/middleware/keep_current_user.rb

```diff
@@ -19,6 +19,10 @@ def finalize
         restore_curent_user { pass }
       end
 
+      def finalize_phase(execution_plan, *args)
+        restore_curent_user(execution_plan.entry_action) { pass(execution_plan, *args) }
```

Comment:
Why are we passing a different action here?

Reply:
The `finalize_phase` method of the middleware is run only once around the entire phase, while regular `finalize` is run around each step in finalize. Because it is run around the entire phase, it doesn't get a specific action as an argument so we just pick the entry action of the execution plan and hope for the best. There are two issues with this approach (entry action not having current user set, each action setting its own current user), but the first one shouldn't happen and the best we can do about the other is strongly discourage people from doing so.

Reply:
What about deprecating it? :shrug:
If we can detect it, we can deprecate and remove in few versions, if noone complains? :thinking:

Reply:
Deprecating what exactly?

Reply:
`user_id` in this middleware. now it is being used to set user for every step, but we can deprecated that, right?

Reply:
But if we keep it as is, it will just override the user for the specific action, right? and if there is no :current_user_id, no user is set, that seems reasonable to me :shrug:

Reply:
> But if we keep it as is, it will just override the user for the specific action, right?

It will, but then the katello issue could get a new twist
1) Finalize phase and transaction starts, user A is set
2) nth step in finalize phase sets its user to B and saves an ActiveRecord object
3) Finalize phase ends, transaction ends, after_commit hooks are executed, but with user A

This is a hypothetical scenario, I don't think it could happen with the action code we currently have.

Reply:
I think it would be enough to document what can happen, but still we can try to detect such scenarios and warn about them, right?

Reply:
In the end I went with a deprecation warning